tools/copyblocks/bucket.go

```diff
@@ -0,0 +1,232 @@
+
+func (bkt *azureBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	client := bkt.containerClient.NewBlobClient(name)
+	response, err := client.DownloadStream(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	return response.Body, nil
+}
+
+func (bkt *azureBucket) Copy(ctx context.Context, name string, dstBucket bucket) error {
+	sourceClient := bkt.containerClient.NewBlobClient(name)
+	sasURL, err := sourceClient.GetSASURL(sas.BlobPermissions{Read: true}, time.Now(), time.Now().Add(10*time.Minute))
+	if err != nil {
+		return err
+	}
+	d, ok := dstBucket.(*azureBucket)
+	if !ok {
+		return errors.New("destination bucket wasn't a blob storage bucket")
+	}
+	dstClient := d.containerClient.NewBlobClient(name)
+
+	response, err := dstClient.StartCopyFromURL(ctx, sasURL, nil)
+	if err != nil {
+		return err
+	}
+
+	copyStatus := response.CopyStatus
+	backoff := backoff.New(ctx, backoff.Config{
+		MinBackoff: time.Second,
+		MaxBackoff: time.Minute,
+		MaxRetries: 10,
+	})
+
+	for {
+		if copyStatus == nil {
+			return errors.New("no copy status present for blob copy")
+		}
+
+		switch *copyStatus {
+		case blob.CopyStatusTypeSuccess:
+			return nil
+		case blob.CopyStatusTypeFailed:
+			return errors.New("copy failed")
+		case blob.CopyStatusTypeAborted:
+			return errors.New("copy aborted")
+		case blob.CopyStatusTypePending:
+			// proceed
+		default:
+			return errors.Errorf("unrecognized copy status: %v", *copyStatus)
+		}
+
+		if !backoff.Ongoing() {
+			break
+		}
+		backoff.Wait()
+
+		response, err := dstClient.GetProperties(ctx, nil)
+		if err != nil {
+			return err
+		}
+		copyStatus = response.CopyStatus
+	}
+	return errors.Wrap(backoff.Err(), "waiting for blob copy status")
+}
+
+func (bkt *azureBucket) ListPrefix(ctx context.Context, prefix string, recursive bool) ([]string, error) {
+	if prefix != "" && !strings.HasSuffix(prefix, delim) {
+		prefix = prefix + delim
+	}
+
+	list := make([]string, 0, 10)
+	if recursive {
+		pager := bkt.containerClient.NewListBlobsFlatPager(&container.ListBlobsFlatOptions{Prefix: &prefix})
+		for pager.More() {
+			page, err := pager.NextPage(ctx)
+			if err != nil {
+				return nil, err
+			}
+			for _, blobItem := range page.Segment.BlobItems {
+				list = append(list, *blobItem.Name)
+			}
+		}
+	} else {
+		pager := bkt.containerClient.NewListBlobsHierarchyPager(delim, &container.ListBlobsHierarchyOptions{Prefix: &prefix})
+		for pager.More() {
+			page, err := pager.NextPage(ctx)
+			if err != nil {
+				return nil, err
+			}
+			for _, blobItem := range page.Segment.BlobItems {
+				list = append(list, *blobItem.Name)
+			}
+			for _, blobPrefix := range page.Segment.BlobPrefixes {
+				list = append(list, *blobPrefix.Name)
+			}
+		}
+	}
+
+	var hasPrefix bool
+	for i, s := range list {
+		list[i], hasPrefix = strings.CutPrefix(s, prefix)
+		if !hasPrefix {
+			return nil, errors.Errorf("listPrefix: path has invalid prefix: %v, expected prefix: %v", s, prefix)
+		}
+	}
```

Comment:
Same question as for GCS: do we ever expect this to happen?

Reply:
Responded to the Azure case. To clarify we don't expect this to happen.